keps/sig-node/4188-kubelet-pod-readiness-api/README.md

```diff
@@ -0,0 +1,343 @@
+
+### Non-Goals
+
+### User Stories
+* Some on node system workloads want to reduce Control Plane dependency and introduce locality for pod’s readiness due to reliability and scalability issues.
+* For building custom monitoring tools that show information about pod’s readiness. When it is ready to serve the traffic.
+
+## Proposal
+
+We are proposing to create a new gRPC API that will return pod's readiness information.
+* The API will return data about both Static and Regular Pods.
+* The API will not return partial data. If kubelet does not know actual information about workloads then appropriate error code will be returned.
+* API should return the most recent information about pods computed by the kubelet even when those data were not reported or accepted by kube-API.
+* The API will be read-only and accessible for on-node workloads (we will use unix socket to connect) without the authorization.
+* The API will be versioned.
+
+
+### Risks and Mitigations
+
+This API is read-only, which removes a large class of risks.
+
+| Risk                                                      | Impact        | Mitigation |
+| --------------------------------------------------------- | ------------- | ---------- |
+| Too many requests to the API impacting the kubelet performances | High          | Rate limiting the API. |
+| Misuse of the API  | High |  This API is Read-only. We will expose only a small portion of the pod's information related to the pod readiness. Exposed data does not contain sensitive information that could be used in a malformed way. |
+| Unauthorized access to the API | Medium | This API is designed to be accessed by all on-node workloads. There is no need to authorize the clients to access this API. |
+| Kubelet restart [issue](https://github.com/kubernetes/kubernetes/issues/100277) | High | This API should serve only complete information about workloads. If kubelet is in the init phase and not all pod’s information is known, then the API should report that information. |
+| Kube-API is down or unreachable | Low | Kube-API availability should not impact this API. When the control plane is down or unreachable but kubelet is working properly this API should return most recent data about local pods that are computed by kubelet even if those data were not reported or accepted by Kube-API. |
+
+#### Control Plane availability issue
+
+The Kubernetes Control Plane is the source of truth and guarantees system integrity. However, already running pods should not be dependent on Control Plane availability or connectivity issues. When Node is healthy but has problems accessing Kube-API running pods, statuses should be refreshed and stored as usual.
+The new API should return pods statuses that are computed by kubelet even though they are not reported back to kube-API but reflect the current pod’s state.
```

Comment:
in a nutshell, this is what I'm kinda concerned: the control plane is the source of truth... but in some cases the kubelet-provided data will be the actual source of truth. Perhaps a step in the right direction would be to make sure that the information is provided by this API only if it is acknowledged by the control plane somehow (e.g. after a succesfull update? that would require firther thought) and so no information misalignment is ever possible?

Reply:
The actual purpose of this GEP is to remove a dependency on control plane availability here. This is coming from a networking perspective where we're trying to understand the health of Pods running locally on the same Node. Since the readiness probes are run locally with Kubelet, it does not make sense for us to have a dependency on control plane for this information. 

Essentially we're trying to avoid a case where we're not updating our routing logic if the control plane is unavailable. This could cause completely unnecessary downtime for applications running on Kubernetes.

Reply:
@ffromani I think @robscott's reply should address your concerns above. Basically the proposed API won't be used to override any decision made by the control plane.